src/middleware.ts

```diff
@@ -258,36 +263,63 @@ export const persist = <S extends State>(
     void setItem()
   }
 
+  const makeThenable = (value: any) => {
+    if (value instanceof Promise) {
+      return value
+    } else {
+      return {
+        then: (callback) => callback(value),
+        // this is a fake function just to avoid catch is not a function error
+        catch: () => undefined,
+      }
+    }
+  }
   // rehydrate initial state with existing stored state
-  ;(async () => {
+  ;(() => {
     const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
 
-    try {
-      const storageValue = await storage.getItem(name)
-
-      if (storageValue) {
-        const deserializedStorageValue = await deserialize(storageValue)
-
-        // if versions mismatch, run migration
-        if (deserializedStorageValue.version !== version) {
-          const migratedState = await migrate?.(
-            deserializedStorageValue.state,
-            deserializedStorageValue.version
-          )
-          if (migratedState) {
-            set(migratedState)
-            await setItem()
+    let storageValuePending = makeThenable(storage.getItem(name))
+    storageValuePending
+      .then((storageValue) => {
+        if (storageValue) {
+          return makeThenable(deserialize(storageValue))
+        } else {
+          return makeThenable(null)
+        }
+      })
+      .then((deserializedStorageValue) => {
+        if (deserializedStorageValue) {
+          if (deserializedStorageValue.version !== version) {
+            return makeThenable(
+              migrate?.(
+                deserializedStorageValue.state,
+                deserializedStorageValue.version
+              )
```

Comment:
I think this needs to be
```suggestion
            
            return makeThenable(
              migrate?.(
                deserializedStorageValue.state,
                deserializedStorageValue.version
              )
              .then((migratedState) => set(migratedState))
              .then(setItem)
```